crates/resolver/src/lookup_state.rs

```diff
@@ -290,19 +290,27 @@ impl<C: DnsHandle + Send + 'static> CachingClient<C> {
                 .filter_map(|r| {
                     // because this resolved potentially recursively, we want the min TTL from the chain
                     let ttl = cname_ttl.min(r.ttl());
-
                     // TODO: disable name validation with ResolverOpts? glibc feature...
                     // restrict to the RData type requested
                     if query.query_class() == r.dns_class() {
                         // standard evaluation, it's an any type or it's the requested type and the search_name matches
-                        // - or -
+                        if (query.query_type().is_any() || query.query_type() == r.rr_type())
```

Comment:
This looks good.

I'm wondering if we should add an option to the always include, or not, and enable it by default only for the ResolverAuthority, but off for backward compatibility?

Adding it to `ResolverOpts` as perhaps a `preserve_intermediates`, which could just hang onto CNAMEs or potentially other future intermediate records? That might give a nice balance between the cost of storing them, vs. the desire to have them for specific scenarios?